### PR TITLE
Add DDnet

### LINKED ISSUE #1093: DDnet
Is this app installable with `sudo apt install`?

Yes but the apt Version is outdated.

If you've created a zip file for this app, upload it here
[DDnet.zip](https://github.com/Botspot/pi-apps/files/7434212/DDnet.zip)

I did not tested it on x64 but im shure it will also work there. I will try it later today or tomorrow. And when it works i will send a ZIP File with the install script for x64 (i think it's the same as for 32x xD)

Here is the newest Version:

https://github.com/Botspot/pi-apps/issues/1093#issuecomment-955682513

### PR DESCRIPTION
**Pull request by theofficialgman**
broken out from PR: https://github.com/Botspot/pi-apps/pull/1239 which is now just DroidCam

this PR changed back to upstream as the forked github repo used is pointless (and now very out of date): https://github.com/KolpixTV/ddnet/commit/4d9167de7a38556375ef8a9e30911c981f06fd58 it has a one string change compared to upstream. `-latomic` should be automatically linked based on the header files... not sure why its not for @KolpixTV. probably an old compiler bug.

needs testing on piOS 32bit (buster or bullseye, but preferably on a pi3 and 4)

and close #1093.﻿

### CODE REVIEW

**Comment by theofficialgman** [comment]
upstream compiled perfectly on arm64. need someone to test on 32bit buster/bullseye I guess

**Comment by Botspot** [comment]
```
--   Found vulkan, version 1.2.162
CMake Error at cmake/BuildVulkanShaders.cmake:25 (message):
  glslangValidator binary was not found.  Did you install the Vulkan SDK /
  packages ?
Call Stack (most recent call first):
  CMakeLists.txt:578 (include)
```
What should be done to fix this error?

**Comment by theofficialgman** [comment]
there are multiple backends, vulkan, opengl, and opengles. we can simply disable vulkan if you don't think its stable on the pi anyway.
I left it enabled since it works on switch but honestly the game is very lite, it doesn't need it.

I do think if you install `glslang-dev` it would go away. at some point I did that manually (since its not available on bionic) and I didn't get this error

**Comment by Botspot** [comment]
Even with `glslang-dev` installed, the same error occured.
```
--   Found vulkan, version 1.2.162
CMake Error at cmake/BuildVulkanShaders.cmake:25 (message):
  glslangValidator binary was not found.  Did you install the Vulkan SDK /
  packages ?
Call Stack (most recent call first):
  CMakeLists.txt:578 (include)


-- Configuring incomplete, errors occurred!
See also "/tmp/ddnet/build/CMakeFiles/CMakeOutput.log".
See also "/tmp/ddnet/build/CMakeFiles/CMakeError.log".
Failed at cmake!
```

**Comment by Botspot** [comment]
Installing `glslang-tools` appears to have fixed this problem. Maybe it should be added as a dependency.

Now there is a new error:
```
-- Configuring done
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
X264_LIBRARY
    linked by target "DDNet" in directory /tmp/ddnet
```

**Comment by theofficialgman** [comment]
looks like another missing dependency, I've added it

File: apps/DDnet/install [hunk 1]
```diff
@@ -1,7 +1,15 @@
 #!/bin/bash
 
 #Dependencies
-install_packages build-essential cmake git libcurl4-openssl-dev libssl-dev libfreetype6-dev libglew-dev libnotify-dev libogg-dev libopus-dev libopusfile-dev libpnglite-dev libsdl2-dev libsqlite3-dev libwavpack-dev python google-mock || exit 1
+install_packages build-essential cmake git libcurl4-openssl-dev libssl-dev libfreetype6-dev libglew-dev libnotify-dev libogg-dev libopus-dev libopusfile-dev libpnglite-dev libsdl2-dev libsqlite3-dev libwavpack-dev python google-mock glslang-tools glslang-dev libx264-dev || exit 1
```
Comment (Botspot) [inline apps/DDnet/install]: I believe you've added glslang-tools glslang-dev dependencies here by mistake.

Comment (theofficialgman) [reply apps/DDnet/install]: I already corrected it before your comment

**Comment by theofficialgman** [comment]
I've tested compiling and running on armhf bionic through my chroot. seems to work

File: apps/DDnet/install [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+
+#Dependencies
+install_packages build-essential cmake git libcurl4-openssl-dev libssl-dev libfreetype6-dev libglew-dev libnotify-dev libogg-dev libopus-dev libopusfile-dev libpnglite-dev libsdl2-dev libsqlite3-dev libwavpack-dev python google-mock libx264-dev || exit 1
+
+package_available glslang-tools
+if [[ $? == "0" ]]; then
+  install_packages glslang-tools glslang-dev || error "Failed to install dependencies"
```
Comment (Botspot) [inline apps/DDnet/install:8]: This seems to be compiling fine for me with glslang-tools installed but glslang-dev not installed.
Once it's done I'll check if it runs.

**Comment by Botspot** [comment]
Other comments:

- This was able to compile and run without glslang-dev being installed. It can probably be safely removed.
- The menu entry has a broken icon. I have not yet checked to see why.
- The description of this app failed to explain the type of game this is. It mentioned racing, so I was imagining something like TrackMania, but this is better described as a platformer game.

**Comment by theofficialgman** [comment]
> * This was able to compile and run without glslang-dev being installed. It can probably be safely removed.

will do

> * The menu entry has a broken icon. I have not yet checked to see why.

yeah its broken for me to... I don't understand why because the icon is written to the icon directory. (the Icon= uses just the name, not a path). This works on other apps... I can't figure out why it doesn't here (maybe it only works for /usr icons ?)

> * The description of this app failed to explain the type of game this is. It mentioned racing, so I was imagining something like TrackMania, but this is better described as a platformer game.

agreed, I kept the description from the original zip which pulled it directly from the website https://ddnet.tw/ but I agree it is not good. the game is a super fun platformer from what I've played of the tutorial.
I'll see if there is a better description or if I can just write my own.

**Comment by theofficialgman** [comment]
the .desktop file comes from upstream. I've opened a bug there incase any .desktop file wizards can give some insight on how that can be fixed without hardcoding https://github.com/ddnet/ddnet/issues/5230

**Comment by Botspot** [comment]
By looking through /tmp/ddnet/build/install_manifest.txt, I found lines pertaining to ddnet.png 
They all appear to go in /usr/local/share. Here is an example line:
`/usr/local/share/icons/hicolor/256x256/apps/ddnet.png`
I checked and it is a valid PNG file.

Perhaps the icon caches were simply not updated?

**Comment by Botspot** [comment]
One other thing:

- This app should be renamed from DDnet to DDNet.

**Comment by Botspot** [comment]
I ran this command:
```
sudo update-icon-caches /usr/local/share/icons/*
```
Nothing happened, so I ran this:
```
sudo xdg-icon-resource forceupdate --mode system
```
Still, nothing happened.
Then I restarted lxpanel with `lxtanelctl restart` and now the icon is working.

I'm not sure which of those commands, if any, were necessary. It could be that if I had restarted the panel first, the icon would have worked.

**Comment by theofficialgman** [comment]
```
sudo update-icon-caches /usr/local/share/icons/*
```
this worked for me immediately in ubuntu budgie
I've done reboots to no avail but this fixed it right away

**Comment by Botspot** [comment]
> ```
> sudo update-icon-caches /usr/local/share/icons/*
> ```
> 
> this worked for me immediately in ubuntu budgie I've done reboots to no avail but this fixed it right away. nice going @Botspot

Good to know.

I tried viewing the list of game servers, and consistently got errors.
```
[2022-05-25 22:58:04][http]: fetching https://info2.ddnet.tw/info?name=nameless%20tee
[2022-05-25 22:58:04][stun/6]: couldn't send stun request
[2022-05-25 22:58:05][stun/4]: got address: 199.195.253.224:26117
[2022-05-25 22:58:05][stun/6]: couldn't send stun request
[2022-05-25 22:58:06][http]: https://master2.ddnet.tw/ddnet/15/servers.json failed. libcurl error: The requested URL returned error: 404 
[2022-05-25 22:58:06][http]: https://master4.ddnet.tw/ddnet/15/servers.json failed. libcurl error: Could not resolve host: master4.ddnet.tw
[2022-05-25 22:58:07][stun/6]: couldn't send stun request
[2022-05-25 22:58:12][stun/6]: couldn't send stun request
[2022-05-25 22:58:20][stun/6]: couldn't send stun request
[2022-05-25 22:58:36][stun/6]: couldn't send stun request
[2022-05-25 22:58:41][http]: https://master1.ddnet.tw/ddnet/15/servers.json failed. libcurl error: Resolving timed out after 10000 milliseconds
[2022-05-25 22:58:41][http]: https://info2.ddnet.tw/info?name=nameless%20tee failed. libcurl error: Resolving timed out after 10000 milliseconds
[2022-05-25 22:58:41][storage]: failed to remove: /home/pi/.local/share/ddnet/ddnet-info.json.19169.tmp
[2022-05-25 22:58:41][http]: https://master3.ddnet.tw/ddnet/15/servers.json failed. libcurl error: Could not resolve host: master3.ddnet.tw
[2022-05-25 22:58:41][serverbrowse_http]: WARNING: no usable masters found
[2022-05-25 22:58:41][serverbrowse_http]: no working serverlist URL found
```
All URLs in the output above are unresolvable 404, at least according to my web browser.

**Comment by theofficialgman** [comment]
hmm... strange. I haven't had any issues previewing game servers. maybe you have some sort of network block?

I was able to connect to master1.ddnet.tw which is online and the server list appears
```
[2022-05-26 00:07:24][http]: fetching https://info2.ddnet.tw/info?name=theofficialgman
[2022-05-26 00:07:24][http]: https://master4.ddnet.tw/ddnet/15/servers.json failed. libcurl error: Could not resolve host: master4.ddnet.tw
[2022-05-26 00:07:24][stun/4]: got address: XXXXXXXXXXXXXX
[2022-05-26 00:07:24][stun/6]: got address: XXXXXXXXXXXXXXXXXXXXXXXX
[2022-05-26 00:07:25][http]: task done https://info2.ddnet.tw/info?name=theofficialgman
[2022-05-26 00:07:25][http]: fetching https://master1.ddnet.tw/ddnet/15/servers.json
[2022-05-26 00:07:26][serverbrowse_http]: found master, url='https://master1.ddnet.tw/ddnet/15/servers.json' time=794ms
[2022-05-26 00:07:26][http]: https://master3.ddnet.tw/ddnet/15/servers.json failed. libcurl error: Could not resolve host: master3.ddnet.tw
[2022-05-26 00:07:26][http]: task done https://master1.ddnet.tw/ddnet/15/servers.json
[2022-05-26 00:07:26][http]: https://master2.ddnet.tw/ddnet/15/servers.json failed. libcurl error: The requested URL returned error: 404 Not Found
[2022-05-26 00:07:26][serverbrowse_http]: determined best master, url='https://master1.ddnet.tw/ddnet/15/servers.json' time=794ms
```

**Comment by Botspot** [comment]
lol - I can't even join the tutorial server anymore.
2022-05-25-230925_1920x1080_scrot

**Comment by theofficialgman** [comment]
btw, I edited your post and deleted the original one, since you have your public IPV4 in it

**Comment by theofficialgman** [comment]
anyway... sound like general internet issues either on your end, your ISPs end, or their end. nothing to worry about

**Comment by Botspot** [comment]
> this to me appears to be a DNS error. thats on your end or on the end of your ISP
> 
> the serverlist is the same for all platforms

I had been messing around with Lokinet, so I wanted a 3rd party (you) to verify if the ddnet servers were down or if it was my fault. Once all sites including github stopped working one by one, I realized that this was either time to reboot, or it was the end of the world.

**Comment by theofficialgman** [comment]
so funny story... somehow I ended up with an x86_64 glslangvalidator binary in my /usr/local/bin folder and that was getting preferred over my aarch64 one in another folder. so whenever I ran the cmake command here (which build the vulkan shaders using glslangvalidator), a ton of box64 instances started running with the x86_64 glslangvalidator shader building

LOL, no wonder so much ram was getting eaten and the vulkan shaders would just be missing

**Comment by theofficialgman** [comment]
@Botspot I added an automatic updater with help from a DDNet dev on discord, pointing me to an official location for the version (since they don't have github releases)

**Comment by Chairn** [comment]
Hello, small dev from DDNet here.

You can disable vulkan with `-DVULKAN=OFF` in cmake calls.
Please note that 32 bits build will 100% crash the game on some maps due to their size. This might depend on the graphic backend you use. OpenGL 1 is usually not crashing as it uses less memory.

**Comment by Botspot** [comment]
> You can disable vulkan with `-DVULKAN=OFF` in cmake calls. Please note that 32 bits build will 100% crash the game on some maps due to their size. This might depend on the graphic backend you use. OpenGL 1 is usually not crashing as it uses less memory.

In your opinion, would OpenGL run more smoothly or more reliably on a Raspberry Pi?

**Comment by C0D3D3V** [comment]
Based on a few articles I found online, I would say for Raspberry Pi 4 is Vulkan faster. But probably not much, I do not know.

**Comment by Chairn** [comment]
I have absolutely no idea, sorry. It is probably better supported but no certitude.

**Comment by Botspot** [comment]
I'm noticing some stability issues with DDNet.
A variety of actions make the program close unexpectedly. There does not appear to be any noticeable errors. One quick way I found to cause a crash is by going to the Tutorial server and pressing P. (periscope)
@theofficialgman, could you try this? If you get the same crash, then this looks to be an issue upstream. Otherwise, it's probably a Vulkan error, and for that reason I'm recompiling DDNet without Vulkan.

**Comment by theofficialgman** [comment]
> I'm noticing some stability issues with DDNet. A variety of actions make the program close unexpectedly. There does not appear to be any noticeable errors. One quick way I found to cause a crash is by going to the Tutorial server and pressing P. (periscope) @theofficialgman, could you try this? If you get the same crash, then this looks to be an issue upstream. Otherwise, it's probably a Vulkan error, and for that reason I'm recompiling DDNet without Vulkan.

No need to recompile, vulkan isn't the default. Opengl 3.0.0 is the default if available. You change the renderer with in game settings.
On arm64 on Linux I have had no crashes or stability issues with all backends. No issues with P from what I remember (don't have access to anything till next week)

**Comment by Botspot** [comment]
Recompiling without Vulkan appears to have fixed it. On-screen health bars also appeared. I didn't see those before.

**Comment by Chairn** [comment]
If you're using master branch, it may be unstable. Better stay on one of the release tag ;)

**Comment by theofficialgman** [comment]
Release tag is used with the merge. Botspot might have been using an older version of the script that built from master

**Comment by theofficialgman** [comment]
Also I'm not bothered if you want to disable vulkan any we ay. It's brand new in this software and I know the pi implementation is sketchy at best...

**Comment by Botspot** [comment]
> Also I'm not bothered if you want to disable vulkan any way. It's brand new in this software and I know the pi implementation is sketchy at best...

It appears that reinstalling DDNet again from the latest script, without disabling Vulkan, still works fine. In the settings, it appears to choose OpenGL 2.1.0 by default.

@theofficialgman, one problem reported to me by @Sussy-OS is that this script installs `libavcodec-extra`, which conflicts with `libavcodec58`.
```
dpkg: libavcodec58:armhf: dependency problems, but removing anyway as you requested:

 vlc-plugin-base:armhf depends on libavcodec58 (>= 7:4.3.3-0+rpt1+deb11u1).

 simplescreenrecorder depends on libavcodec58 (>= 7:4.2).

 omxplayer depends on libavcodec58 (>= 7:4.2).

 libopenjfx-jni depends on libavcodec58 (>= 7:4.2).

 libchromaprint1:armhf depends on libavcodec58 (>= 7:4.0).

 libavformat58:armhf depends on libavcodec58 (= 7:4.3.4-0+deb11u1+rpt1).

 libavfilter7:armhf depends on libavcodec58 (= 7:4.3.4-0+deb11u1+rpt1).

 libavdevice58:armhf depends on libavcodec58 (= 7:4.3.4-0+deb11u1+rpt1).

 libaubio5:armhf depends on libavcodec58 (>= 7:4.2).

 libasound2-plugins:armhf depends on libavcodec58 (>= 7:4.2).

 gstreamer1.0-libav:armhf depends on libavcodec58 (>= 7:4.2).

 ffmpeg depends on libavcodec58 (= 7:4.3.4-0+deb11u1+rpt1).
```

**Comment by theofficialgman** [comment]
ugh... `libavcodec-extra` requires `libavcodec-extra58` which should be an alternative for `libavcodec-58`

I don't think the extras are needed, I'll just revert to `libavcodec-dev` which won't cause any conflicts no matter what is installed

**Pull request merged**